xarray/backends/api.py

```diff
@@ -1827,18 +1832,247 @@ def to_zarr(
+        raise ValueError("The ``compute`` kwarg is not supported in `initialize_zarr`.")
+
+    if not any(is_chunked_array(v._data) for v in ds._variables.values()):
+        raise ValueError("This function should be used with chunked Datasets.")
+
+    if mode not in ["w", "w-"]:
+        raise ValueError(
+            f"Only mode='w' or mode='w-' is allowed for initialize_zarr. Received mode={mode!r}"
+        )
+
+    if zarr_version is None:
+        # default to 2 if store doesn't specify it's version (e.g. a path)
+        zarr_version = int(getattr(store, "_store_version", 2))
+
+    # The design here is to write to an in-memory temporary store,
+    # and flush that to the actual `store`. This is a major improvement
+    # for V3 high-latency stores (e.g. cloud buckets)
+    if zarr_version == 2:
+        temp_store = zarr.MemoryStore()
+    elif zarr_version == 3:
+        temp_store = zarr.MemoryStoreV3()
+        if consolidated:
+            raise ValueError("Consolidating metadata is not supported by Zarr V3.")
+    else:
+        raise ValueError(f"Invalid zarr_version={zarr_version}.")
+
+    # Needed to handle `store` being a string path
+    store = zarr.hierarchy.normalize_store_arg(
+        store,
+        zarr_version=zarr_version,
+        mode=mode,
+        storage_options=kwargs.get("storage_options", None),
+    )
+    if mode == "w-":
+        # assert that the path does not already exist
+        zarr.open_group(
+            store,
+            mode=mode,
+            storage_options=kwargs.get("storage_options", None),
+            path=kwargs.get("group", None),
+        )
+
+    # Use this to open the group once with all the expected default options
+    # We will reuse xzstore.zarr_group later.
+    xtempstore = backends.ZarrStore.open_group(
+        temp_store,
+        mode="w",  # always write to the temp store
+        zarr_version=zarr_version,
+        consolidated=False,
+        **kwargs,
+    )
+
+    # need to do this separately to get the "coordinates" attribute coorect
+    variables, attrs = conventions.encode_dataset_coordinates(ds)
+    ds = ds._replace(variables=variables, attrs=attrs)
+
+    all_variables = set(ds._variables)
+    # TODO: how do we work with the new index API?
+    index_vars = {dim for dim in ds.dims if dim in all_variables}
+    vars_without_region = {
+        key
+        for key in all_variables - index_vars
+        if (not (set(ds._variables[key].dims) & set(region_dims)))
+    }
+    chunked_vars_without_region = {
+        key for key in vars_without_region if is_chunked_array(ds._variables[key])
+    }
+
+    def extract_encoding(varnames: Iterable[Hashable]) -> dict:
+        return {k: v for k, v in encoding.items() if k in varnames}
+
+    # Always write index variables, and any in-memory variables without region dims
+    eager_write_vars = index_vars | (vars_without_region - chunked_vars_without_region)
+
+    with warnings.catch_warnings():
```

Comment:
Alternatively this function could just duplicate all steps in `dump_to_store`. I'm not sure which is preferable.